### DISCUSSION TITLE
Best way to promisify state machines

### DISCUSSION THREAD

**Discussion by zacharyweidenbach** [statelyai/xstate#3604] [Q&A] [answered]
In the past, I _think_ a running state machine (`const someMachineService = interpret(someMachine)`) exposed a `waitFor` or `wait` method that could be used to promisify the state machine moving from one state to another. But this was removed, I imagine for the reason that there can be a lot of edge cases, like timeouts. So I'm wondering if there is a best practice way to achieve this.

Take for instance, a machine that starts in an `rehydrating` state. When it is started via `someMachineService.start()`, it moves from `rehydrating` -> `ready`. But on the `rehydrating` state, it immediately invokes an async service function that gets some data out of some persistent storage and store it in the machine's context so that it is synchronously available.

So if I wanted to promisify the interface to start the machine service, it might look something like this:

```typescript
const start = () =>
  new Promise((resolve) => {
    const listener = someMachineService.subscribe((state) => {
      if (state.changed && state.matches('ready')) {
        resolve({});
        listener.unsubscribe();
      }
    });
    someMachineService.start();
  });
```

But there are some problems with this. It's not handling any error of course. But also there is no timeout if it never moves to the next state. The latter could be solved easily enough with a setTimeout, or perhaps there is a way in the machine config to abandon the invoked service after a certain amount of time via delays. The example approach I've provided also uses the `subscribe` interface to listen for changes, and then clean up after itself via `unsubscribe()`. Would it be better to listen for the machine to emit some kind of `DONE_REHYDRATING` event into an event bus when it is done transitioning, instead of looking at the machine's current state?

**Comment by with-heart** [comment, marked as answer]
Not sure if `waitFor` alone would resolve your scenario, but it still exists under `xstate/lib/waitFor`

**Reply by davidkpiano** [reply]
Yes, `waitFor` would be the way to "promisify" a state machine:

https://xstate.js.org/docs/guides/interpretation.html#waitfor
